src/Core/QueryProcessingStage.h

```diff
@@ -12,15 +12,29 @@ namespace QueryProcessingStage
     /// Numbers matter - the later stage has a larger number.
     enum Enum
     {
-        FetchColumns       = 0,    /// Only read/have been read the columns specified in the query.
-        WithMergeableState = 1,    /// Until the stage where the results of processing on different servers can be combined.
-        Complete           = 2,    /// Completely.
+        /// Only read/have been read the columns specified in the query.
+        FetchColumns       = 0,
+        /// Until the stage where the results of processing on different servers can be combined.
+        WithMergeableState = 1,
+        /// Until the stage where the aggregate functions were calculated and finalized.
+        ///
+        /// It is used for auto distributed_group_by_no_merge optimization for distributed engine.
+        /// (See comments in StorageDistributed).
+        WithMergeableStateAfterAggregation = 2,
+        /// Completely.
+        Complete           = 3,
     };
 
     inline const char * toString(UInt64 stage)
     {
-        static const char * data[] = { "FetchColumns", "WithMergeableState", "Complete" };
-        return stage < 3
+        static const char * data[] =
+        {
```

Comment:
This is part of protocol, hence values cannot be inserted in the middle, breaks ABI